pkg/agent/client/experimental.go

```diff
@@ -0,0 +1,357 @@
+				telemetry.Error:          err.Error(),
+			}).Warn("Received malformed entry from SPIRE server")
+			continue
+		}
+
+		// Get all federated trust domains
+		for _, td := range entry.FederatesWith {
+			federatesWith[td] = true
+		}
+		regEntries[entry.EntryId] = entry
+	}
+
+	keys := make([]string, 0, len(federatesWith))
+	for key := range federatesWith {
+		keys = append(keys, key)
+	}
+
+	protoBundles, err := c.fetchBundles(ctx, keys)
+	if err != nil {
+		return nil, err
+	}
+
+	bundles := make(map[string]*common.Bundle)
+	for _, b := range protoBundles {
+		bundle, err := bundleFormProto(b)
+		if err != nil {
+			c.c.Log.WithError(err).WithField(telemetry.TrustDomainID, b.TrustDomain).Warn("Received malformed entry from SPIRE server")
+		}
+		bundles[b.TrustDomain] = bundle
+	}
+
+	svids := make(map[string]*node.X509SVID)
+	if len(req.Csrs) > 0 {
+		var params []*svidpb.NewX509SVIDParams
+		for entryID, csr := range req.Csrs {
+			params = append(params, &svidpb.NewX509SVIDParams{
+				EntryId: entryID,
+				Csr:     csr,
+			})
+		}
+
+		protoSvids, err := c.fetchSVIDs(ctx, params)
+		if err != nil {
+			return nil, err
+		}
+
+		for i, s := range protoSvids {
+			entryID := params[i].EntryId
+			if s == nil {
+				c.c.Log.WithField(telemetry.RegistrationID, entryID).Debug("Entry not found")
+				continue
+			}
+			var certChain []byte
+			for _, cert := range s.CertChain {
+				certChain = append(certChain, cert...)
+			}
+
+			svids[entryID] = &node.X509SVID{
+				CertChain: certChain,
+				ExpiresAt: s.ExpiresAt,
+			}
+		}
+	}
+
+	return &Update{
+		Entries: regEntries,
+		Bundles: bundles,
+		SVIDs:   svids,
+	}, nil
+}
+
+func (c *client) fetchEntries(ctx context.Context) ([]*types.Entry, error) {
+	entryClient, connection, err := c.newEntryClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer connection.Release()
+
+	resp, err := entryClient.GetAuthorizedEntries(ctx, &entrypb.GetAuthorizedEntriesRequest{})
+	if err != nil {
+		c.release(connection)
+		c.c.Log.WithError(err).Error("Failed to fetch authorized entries")
+		return nil, errors.New("failed to fetch authorized entries")
+	}
+
+	return resp.Entries, err
+}
+
+func (c *client) fetchBundles(ctx context.Context, federatedBundles []string) ([]*types.Bundle, error) {
+	bundleClient, connection, err := c.newBundleClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer connection.Release()
+
+	var bundles []*types.Bundle
+
+	// Get bundle
+	bundle, err := bundleClient.GetBundle(ctx, &bundlepb.GetBundleRequest{})
+	if err != nil {
+		c.release(connection)
+		c.c.Log.WithError(err).Error("Failed to fetch bundle")
+		return nil, errors.New("failed to fetch bundle")
+	}
+	bundles = append(bundles, bundle)
+
+	for _, b := range federatedBundles {
+		bundle, err := bundleClient.GetFederatedBundle(ctx, &bundlepb.GetFederatedBundleRequest{
+			TrustDomain: b,
+		})
+		if err != nil {
+			// TODO: may it continue or just return error?
```

Comment:
IMO we should just fail this and return the error. I think that it would be difficult to debug the problem otherwise.
@azdagron what's your feeling?

Reply:
Considering this function is also in charge of keeping the trust domain bundle in sync as well as the federated bundles, i'd hesitate to fail the whole function because of a problem fetching a single bundle. I'd also under the normal course of federation for this RPC to return `NotFound` when a federated bundle is removed (and entries) while the agent is in the middle of this loop and hasn't had a chance to get the updated entries.

Maybe the right compromise is to ignore or warn on `NotFound` errors but otherwise continue and then fail on other kinds of errors?

Reply:
I just pushed code to validates if a not found is returned and warn in that case, or fail if not